### DISCUSSION TITLE
Download speed: website vs CLI

### DISCUSSION THREAD

**Discussion by rly** [dandi/helpdesk#24] [Q&A] [answered]
For downloading a single NWB file from DANDI, is there any difference in speed between downloading from the DANDI website versus using the CLI?

**Comment by yarikoptic** [comment, marked as answer]
Shouldn't be AFAIK. IIRC multithread uploads across chunks, but for download I think we aren't doing that, so should be the same speed. Although who knows if browser wouldn't try to chunk it up.
There might be a small difference though for small files since iirc CLI would first inquire details from API server, butt for benefit of doing checksuming while downloading iirc ;-)
Do you observe differences?

**Reply by rly** [reply]
Thanks. I have not observed differences but I will test that. 

I was just wondering if there was a way to speed up a file download. I don't think it is my internet (84 GB at 200 Mbps should take about 1 hour, but is instead taking 4 hours).

**Reply by waxlamp** [reply]
I think in both cases, the actual download happens via an S3 URL. I think that if you're not saturating your pipe, it's likely a lack of bandwidth on Amazon's end (though this is just me speculating).

**Reply by yarikoptic** [reply]
In my experience -- to really saturate the pipe you might want to multithread that download (via range requests) of a single file, or download/upload multiple files at once.  
You could try aria2c and instruct it to go multiple threads, although in my experience I got "varying results" with that.  Here is some s3 related notes (but to a specific host) comparing wget and aria2c:
```
  from S3:
  (https://dandiarchive.s3.amazonaws.com/blobs/9e2/ad0/9e2ad05f-4bb1-4bc3-8874-f4ca2de5edd0)
  wget speeds up slowly from 20MBps to 50MBps and more, averaging at 80
  MBps, on rerun -- saturates at moments 106MBps and averages 103 MBps.

  aria2c without any options seems to suck big time at 12 MiB ....
  yikes... even with -j 6 --split 6 -- takes longer to speed up to 90
  MiB... odd
```
I have tried again to smaug (on a 10Gb pipe. I saw upload/download speeds with S3 in 100s of MBps) wget first run `2.14G  71.6MB/s    in 83s ` although oddly speeds reported while download was ongoing were slowly growing from 3 to 30 MB/s -- might have missed when it became fast; redownload --  was going wobbling in 3-10 MB/s, ended up  `12.7MB/s    in 4m 39s` (so that last MB/s is the last one to be observed, not average).

```
$> aria2c --split 5 -j 5 https://dandiarchive.s3.amazonaws.com/blobs/9e2/ad0/9e2ad05f-4bb1-4bc3-8874-f4ca2de5edd0

09/01 18:01:58 [NOTICE] Downloading 1 item(s)
[#b0487d 2.0GiB/2.1GiB(96%) CN:1 DL:83MiB]                                                               
09/01 18:02:32 [NOTICE] Download complete: /mnt/scrap/tmp/9e2ad05f-4bb1-4bc3-8874-f4ca2de5edd0

Download Results:
gid   |stat|avg speed  |path/URI
======+====+===========+=======================================================
b0487d|OK  |    65MiB/s|/mnt/scrap/tmp/9e2ad05f-4bb1-4bc3-8874-f4ca2de5edd0

Status Legend:
(OK):download completed.
aria2c --split 5 -j 5   4.43s user 5.48s system 29% cpu 33.665 total
```
so it did better!  Using datalad to prallelize multiple downloads at once (`datalad install -J 5 -g https://github.com/dandisets/000003`) I guess has a bug since does not update real time "Total" speed, but each file is getting it at ~10  MB/s, so the same avg speed per file, but 5 times more in total. Running at -J10, slows down to a wobble 3-9 MBps.  So - feels that pipe is saturated somewhere allowing only for up to about 30-90 MBps for me ATM, seems matching the aria2c performance.

**Reply by rly** [reply]
Interesting. Thanks for running some tests. I hadn't used aria2c before but on my machine and network, it doesn't seem to be any faster than wget for a single file (2-10 MB/s). In contrast, downloading from Google Drive is a bit faster at 15-30 MB/s.

Downloading using the dandi CLI seems to take about the same time as downloading from the website via browser.

**Reply by yarikoptic** [reply]
Probably S3 zone for our bucket is also relevant here ;-)
